Remove whitespace: pre because of extra empty lines bug

aldo-expensify opened the pull request:
### Details
This PR needs the changes from the PR: https://github.com/Expensify/Web-Expensify/pull/33993 . This PR without PR#33993 shouldn't break anything.

This PR is reverting partially what was done here: https://github.com/Expensify/App/pull/9025/files

There are email clients like outlook that add a bunch of `\n` in between the html of the email. The `whitespace: pre` cause these `\n` to be rendered as empty lines.

This PR is applying `whitespace: pre` to messages that don't have `source: "email"` and `whitespace: normal` for messages that have it. Email clients don't use `whitespace: pre` when showing the email body, they use `&nbsp` to force/keep whitespaces.

When testing / QA we should make sure that the test / QA from these PRs also passes:

- https://github.com/Expensify/App/pull/9025
- https://github.com/Expensify/App/pull/6975


### Fixed Issues

$ https://github.com/Expensify/App/issues/8088

# Tests

1. Create two accounts, one using a free email from https://outlook.live.com/
2. Log in into NewDot with one of these accounts and create a chat with the other, save the chat ID: `http://localhost:8080/r/<chat ID>`
3. Send a comment with the following content (without codeblock):
    <pre>
   0 spaces
    1 spaces
     2 spaces
      3 spaces
       4 spaces
    </pre>
4. Send another comment with the following content (including a code block)
    <pre>
   0 spaces
    1 spaces
     2 spaces
      3 spaces
       4 spaces

   ```
   some code
   ```
    </pre>
5. You should see that whitespaces are respected and that there aren't extra white lines added
    <img width="535" alt="image" src="https://user-images.githubusercontent.com/87341702/174684136-bbd99830-b132-40a6-8723-4bf8631a2a0b.png">
7. Send an email from the outlook email account so it gets converted as a message in the chat you just created. This [SO](https://stackoverflow.com/c/expensify/a/5750/419) is useful to get that going in DEV. You should be sending an email with:
    - To: `"Expensify rID:<replace-chat-id-here>" <postmaster@sandboxbdd8eb86efaa4dd68b56e83b7b379bd9.mailgun.org>`
    - Subject: What ever you configured (for example, in my case I use "ALDO_TESTING_CREATE_REPORT_COMMENT")
    - Body: type some message with some line jumps like this:
        <img width="1006" alt="image" src="https://user-images.githubusercontent.com/87341702/174683926-3214847f-3e09-4c0e-b890-4b2405cc1465.png">
8. The email should appear in the new dot conversation
     <img width="527" alt="image" src="https://user-images.githubusercontent.com/87341702/174683865-0ad7d82b-bfba-4e05-97f9-56d1bd54a330.png">
5. You should see that it doesn't have an excessive amount of empty lines


# QA (Production)

1. Create two accounts, one using a free email from https://outlook.live.com/
2. Log in into NewDot with one of these accounts and create a chat with the other, save the chat ID: `https://new.expensify.com/r/<chat ID>`
3. Send a comment with the following content (without codeblock):
    <pre>
   0 spaces
    1 spaces
     2 spaces
      3 spaces
       4 spaces
    </pre>
4. Send another comment with the following content (including a code block)
    <pre>
   0 spaces
    1 spaces
     2 spaces
      3 spaces
       4 spaces

   ```
   some code
   ```
    </pre>
5. You should see that whitespaces are respected and that there aren't extra white lines added
    <img width="535" alt="image" src="https://user-images.githubusercontent.com/87341702/174684136-bbd99830-b132-40a6-8723-4bf8631a2a0b.png">
8. Send an email from the outlook email account so it gets converted as a message in the chat you just created. You should be sending an email with:
    - To: `replies+<replace-your-chat-id-here>@expensify.com`
    - Subject: Anything
    - Body: type some message with some line jumps like this:
        <img width="694" alt="Screen Shot 2022-06-20 at 3 01 58 PM" src="https://user-images.githubusercontent.com/87341702/174683998-e9d5a8d6-3d32-4403-a903-9a102d67c4dd.png">
9. The email should appear in the new dot conversation
     <img width="527" alt="image" src="https://user-images.githubusercontent.com/87341702/174683865-0ad7d82b-bfba-4e05-97f9-56d1bd54a330.png">
5. You should see that it doesn't have an excessive amount of empty lines



- [ ] Verify that no errors appear in the JS console

### PR Review Checklist
<!--
This is a checklist for PR authors & reviewers. Please make sure to complete all tasks and check them off once you do, or else Expensify has the right not to merge your PR!
-->
#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I ran the tests on **all platforms** & verified they passed on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [x] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.


<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

- [ ] Verify that no errors appear in the JS console

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the desktop platform-->
<table>
<tr><th>Before fix</th><th>After fix</th></tr>
<tr>
<td>
<img width="400" alt="image" src="https://user-images.githubusercontent.com/87341702/170595024-100a9ee2-979c-44cf-8bbd-0e6a1849216e.png">
</td>
<td>
<img width="400" alt="image" src="https://user-images.githubusercontent.com/87341702/170595276-ee25c817-2911-407f-9fc6-83ccd9e03d7d.png">
</td>
</tr>
<tr>
<td>
<img width="400" alt="image" src="https://user-images.githubusercontent.com/87341702/170591130-590a80c8-45d3-4cc5-b4d3-7f2a2b8ed917.png">
</td>
<td>
<img width="400" alt="image" src="https://user-images.githubusercontent.com/87341702/170591239-05307dc6-e5b5-412c-aa74-a8ce0da12df8.png">
</td>
</tr>
</table>

#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->

#### Desktop


#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

#### Android
<!-- Insert screenshots of your changes on the Android platform-->
<table>
<tr><th>Before fix</th><th>After fix</th></tr>
<tr>
<td>
<img width="400" alt="image" src="https://user-images.githubusercontent.com/87341702/170595421-5ccfd219-4d79-49fd-ac99-081e9e59bc62.png">
</td>
<td>
<img width="400" alt="image" src="https://user-images.githubusercontent.com/87341702/170595416-7c6d5a3d-c7bc-4d03-a4a7-57c3efa7a0a3.png">
</td>
</tr>
<tr>
<td>
<img width="400" alt="image" src="https://user-images.githubusercontent.com/87341702/170591443-f1e0554a-f2d7-42e5-afa7-5fa3a9157c9d.png">
</td>
<td>
<img width="400" alt="image" src="https://user-images.githubusercontent.com/87341702/170591471-5ba627a0-3fd2-467f-b4d4-fdd257254046.png">
</td>
</tr>
</table>

aldo-expensify committed:
Remove whitespace: pre because extra empty lines bug

aldo-expensify commented:
@parasharrajat since you are somewhat familiarized with the problem... do you know of any regressions this PR could cause by removing `whiteSpace: 'pre',`?

aldo-expensify commented:
I can see that this breaks in some cases when you copy/paste HTML stuff that had white spaces. I.e. copy the sample text from test step 3) and paste it as comment using `CMD + SHIFT + V`, whitespaces will be lost:

<img width="364" alt="Screen Shot 2022-05-26 at 7 47 43 PM" src="https://user-images.githubusercontent.com/87341702/170601772-44961fd7-6177-47db-96d3-6894f819cb1a.png">

The whitespaces are simply gone when looking at the DOM. The spaces seem to be getting lost in the rendering process because I can see them coming from the backend in the Report_History:

```
"0 spaces<br /> 1 spaces<br />  2 spaces<br />   3 spaces<br />    4 spaces<br /> <pre>some code</pre><pre>some more      code<br /><br />asdad asd asd. asda.       asd</pre><br />1. Some item<br />    i. Some subitem 1<br />    ii. Some subitem 2<br />2. Some item 2"
```

And this is one that looks fine:

```
"0 space<br />  2 spaces<br />    4 spaces<br />      6 spaces"
```
<img width="136" alt="Screen Shot 2022-05-26 at 7 52 16 PM" src="https://user-images.githubusercontent.com/87341702/170602151-41a7ba69-c506-49e3-9205-65642a5ecd16.png">

I'm not sure about what is the difference yet.

parasharrajat commented:
I will dig this over the weekend and let you know. Thanks for waiting.

aldo-expensify committed:
Merge branch 'aldo_fix-empty-lines' of github.com:Expensify/App into aldo_fix-empty-lines

aldo-expensify commented:
The easiest way to reproduce is to add this line in the promise handler for `DeprecatedAPI.Report_GetHistory` [here](https://github.com/Expensify/App/blob/50da22fa41944f42b7c36ba7a91f0ab737eb6cc6/src/libs/actions/Report.js#L945):

<img width="765" alt="image" src="https://user-images.githubusercontent.com/87341702/171970891-d742d1d0-35fb-486a-878d-e6799a39bd4b.png">


```js
            data.history = [{"person":[{"type":"TEXT","style":"strong","text":"aldo+test-spacing1@expensifail.com"}],"actorEmail":"aldo+test-spacing1@expensifail.com","actorAccountID":11030556,"message":[{"type":"COMMENT","html":"     <div class=\"WordSection1\"> <p class=\"MsoNormal\">Hi SomeName, <\/p><p><\/p> <p class=\"MsoNormal\"><\/p><p>\u00a0<\/p> <p class=\"MsoNormal\">Will tomorrow at 34 est. work?<\/p><p><\/p> <p class=\"MsoNormal\"><\/p><p>\u00a0<\/p> <p class=\"MsoNormal\">Thanks,  Test<\/p><p><\/p> <p class=\"MsoNormal\"><\/p><p>\u00a0<\/p> <\/div>xxxxxx xxxxxxxx xx xxx: xxxx x-xxxx xxx xxxxxxxxxx xx xxxxxxxxxxxxx xx xxxxxxxxxxxx xxxxx x.x. xxx, xx xxx xxxxxxx xxxxxxx xx xx xxxxxxxxx xx xxxxxxx x xxxxxxxxxx xxxxxxx xx xxxxxxx. xxx xxx xxxxxx xxx xx xxxxxxx xxxxxxxxxxx xxx xxxxxxxxxxx xxxxxxxx xxxx xxxxx  xxx (xxxxxx xxx xxx xxxxx.xxx xxxxxxx, xxxxx xxxxxx x-xxxx xxxxxxxxxxx xxxxxxx x xxxxxxxx xxxxxxx) xx xxxx x-xxxx xxxxxxx xx xxxxxxxxxx xxxx xxxxxxx xx xxxxxxxxxxxxxxxx@xxxxx.xxx. xx xxx xx xx, xxx xxxxxx xx xxxx xxxxxxx xxxx xx xxxxxxxx xxxxxxxx. xxx xxxxxxxxx  xxxxxx xxxxxxx xx 225 x xxxxxx xxxxx, xxxxx 2600, xxxxxxx, xx 60602-4903.  xxxx xxxxx xxxxxxx xx xxxx xxxxx xxx xx xxx xx xxx xxxxxxxxxxxx xxx xxx xxxxxxx xxxxxxxxxx xx xxxxxxxxxxxx xxxxxxxxxxx  xx xxxxx xxxxxxxxxxx xxxxxx xxxx xxxxxxxxxx xxxxx xxxxxxxxxx xxx. xx xxx xxx xxx xxx xxxxxxxx xxxxxxxxx, xxxxxx xxxxxx xxx xxxxxx xx xxxxx xxxxx xxxxxxxxxxx xxx xxxxxx xxxx xxxxxxx xxxxxxx xxxxxxx xxxxxxx xx xxxxxxxxxx xx xxxxxx. xxxx xxxxx xx xxx xxxxxxxx  xx xx x xxxxxxxx xx xxxxx xxxxxxx xxxxxxx xxxxxxxxxx, xxx xxx xxx xxxxxx xxxxxxxxx xx xxxx xxxxx xxxxxx xxx xx xxxxxxxxx xx x xxxxxx xxx xxxxxxx xxxxxx xxxxxxxxx xxxxxx. xxxxx xxx.xxxxx.xxx\/xxxxxxxxxx xxx xxxx xxxxxxxxxxx xxxxx xxxxx xxx xxx xxx xxxxxxxxxxxx.     ","text":"      Hi SomeName,  \u00a0 Will tomorrow at 34 est. work? \u00a0 Thanks,  Test \u00a0 xxxxxx xxxxxxxx xx xxx: xxxx x-xxxx xxx xxxxxxxxxx xx xxxxxxxxxxxxx xx xxxxxxxxxxxx xxxxx x.x. xxx, xx xxx xxxxxxx xxxxxxx xx xx xxxxxxxxx xx xxxxxxx x xxxxxxxxxx xxxxxxx xx xxxxxxx. xxx xxx xxxxxx xxx xx xxxxxxx xxxxxxxxxxx xxx xxxxxxxxxxx xxxxxxxx xxxx xxxxx  xxx (xxxxxx xxx xxx xxxxx.xxx xxxxxxx, xxxxx xxxxxx x-xxxx xxxxxxxxxxx xxxxxxx x xxxxxxxx xxxxxxx) xx xxxx x-xxxx xxxxxxx xx xxxxxxxxxx xxxx xxxxxxx xx xxxxxxxxxxxxxxxx@xxxxx.xxx. xx xxx xx xx, xxx xxxxxx xx xxxx xxxxxxx xxxx xx xxxxxxxx xxxxxxxx. xxx xxxxxxxxx  xxxxxx xxxxxxx xx 225 x xxxxxx xxxxx, xxxxx 2600, xxxxxxx, xx 60602-4903.  xxxx xxxxx xxxxxxx xx xxxx xxxxx xxx xx xxx xx xxx xxxxxxxxxxxx xxx xxx xxxxxxx xxxxxxxxxx xx xxxxxxxxxxxx xxxxxxxxxxx  xx xxxxx xxxxxxxxxxx xxxxxx xxxx xxxxxxxxxx xxxxx xxxxxxxxxx xxx. xx xxx xxx xxx xxx xxxxxxxx xxxxxxxxx, xxxxxx xxxxxx xxx xxxxxx xx xxxxx xxxxx xxxxxxxxxxx xxx xxxxxx xxxx xxxxxxx xxxxxxx xxxxxxx xxxxxxx xx xxxxxxxxxx xx xxxxxx. xxxx xxxxx xx xxx xxxxxxxx  xx xx x xxxxxxxx xx xxxxx xxxxxxx xxxxxxx xxxxxxxxxx, xxx xxx xxx xxxxxx xxxxxxxxx xx xxxx xxxxx xxxxxx xxx xx xxxxxxxxx xx x xxxxxx xxx xxxxxxx xxxxxx xxxxxxxxx xxxxxx. xxxxx xxx.xxxxx.xxx\/xxxxxxxxxx xxx xxxx xxxxxxxxxxx xxxxx xxxxx xxx xxx xxx xxxxxxxxxxxx.     ","isEdited":false}],"originalMessage":{"html":"     <div class=\"WordSection1\"> <p class=\"MsoNormal\">Hi SomeName, <\/p><p><\/p> <p class=\"MsoNormal\"><\/p><p>\u00a0<\/p> <p class=\"MsoNormal\">Will tomorrow at 34 est. work?<\/p><p><\/p> <p class=\"MsoNormal\"><\/p><p>\u00a0<\/p> <p class=\"MsoNormal\">Thanks,  Test<\/p><p><\/p> <p class=\"MsoNormal\"><\/p><p>\u00a0<\/p> <\/div>xxxxxx xxxxxxxx xx xxx: xxxx x-xxxx xxx xxxxxxxxxx xx xxxxxxxxxxxxx xx xxxxxxxxxxxx xxxxx x.x. xxx, xx xxx xxxxxxx xxxxxxx xx xx xxxxxxxxx xx xxxxxxx x xxxxxxxxxx xxxxxxx xx xxxxxxx. xxx xxx xxxxxx xxx xx xxxxxxx xxxxxxxxxxx xxx xxxxxxxxxxx xxxxxxxx xxxx xxxxx  xxx (xxxxxx xxx xxx xxxxx.xxx xxxxxxx, xxxxx xxxxxx x-xxxx xxxxxxxxxxx xxxxxxx x xxxxxxxx xxxxxxx) xx xxxx x-xxxx xxxxxxx xx xxxxxxxxxx xxxx xxxxxxx xx xxxxxxxxxxxxxxxx@xxxxx.xxx. xx xxx xx xx, xxx xxxxxx xx xxxx xxxxxxx xxxx xx xxxxxxxx xxxxxxxx. xxx xxxxxxxxx  xxxxxx xxxxxxx xx 225 x xxxxxx xxxxx, xxxxx 2600, xxxxxxx, xx 60602-4903.  xxxx xxxxx xxxxxxx xx xxxx xxxxx xxx xx xxx xx xxx xxxxxxxxxxxx xxx xxx xxxxxxx xxxxxxxxxx xx xxxxxxxxxxxx xxxxxxxxxxx  xx xxxxx xxxxxxxxxxx xxxxxx xxxx xxxxxxxxxx xxxxx xxxxxxxxxx xxx. xx xxx xxx xxx xxx xxxxxxxx xxxxxxxxx, xxxxxx xxxxxx xxx xxxxxx xx xxxxx xxxxx xxxxxxxxxxx xxx xxxxxx xxxx xxxxxxx xxxxxxx xxxxxxx xxxxxxx xx xxxxxxxxxx xx xxxxxx. xxxx xxxxx xx xxx xxxxxxxx  xx xx x xxxxxxxx xx xxxxx xxxxxxx xxxxxxx xxxxxxxxxx, xxx xxx xxx xxxxxx xxxxxxxxx xx xxxx xxxxx xxxxxx xxx xx xxxxxxxxx xx x xxxxxx xxx xxxxxxx xxxxxx xxxxxxxxx xxxxxx. xxxxx xxx.xxxxx.xxx\/xxxxxxxxxx xxx xxxx xxxxxxxxxxx xxxxx xxxxx xxx xxx xxx xxxxxxxxxxxx.     "},"avatar":"https:\/\/d2k5nsl2zxldvw.cloudfront.net\/images\/avatars\/avatar_2.png","created":"May 28 2022 9:10am PDT","timestamp":1653754210,"automatic":false,"sequenceNumber":3,"actionName":"ADDCOMMENT","shouldShow":true,"clientID":"","reportActionID":"63777"},{"person":[{"type":"TEXT","style":"strong","text":"aldo+test-spacing1@expensifail.com"}],"actorEmail":"aldo+test-spacing1@expensifail.com","actorAccountID":11030556,"message":[{"type":"COMMENT","html":"0 spaces<br \/> 1 spaces<br \/>  2 spaces<br \/>   3 spaces<br \/>    4 spaces<br \/> <pre>Test<\/pre>","text":"0 spaces  1 spaces   2 spaces    3 spaces     4 spaces  Test","isEdited":false}],"originalMessage":{"clientID":"1654303872951187","html":"0 spaces<br \/> 1 spaces<br \/>  2 spaces<br \/>   3 spaces<br \/>    4 spaces<br \/> <pre>Test<\/pre>"},"avatar":"https:\/\/d2k5nsl2zxldvw.cloudfront.net\/images\/avatars\/avatar_2.png","created":"Jun 3 2022 5:51pm PDT","timestamp":1654303873,"automatic":false,"sequenceNumber":2,"actionName":"ADDCOMMENT","shouldShow":true,"clientID":"1654303872951187","reportActionID":"63776"},{"person":[{"type":"TEXT","style":"strong","text":"aldo+test-spacing1@expensifail.com"}],"actorEmail":"aldo+test-spacing1@expensifail.com","actorAccountID":11030556,"message":[{"type":"COMMENT","html":"0 spaces<br \/> 1 spaces<br \/>  2 spaces<br \/>   3 spaces<br \/>    4 spaces","text":"0 spaces  1 spaces   2 spaces    3 spaces     4 spaces","isEdited":false}],"originalMessage":{"clientID":"1654303847866358","html":"0 spaces<br \/> 1 spaces<br \/>  2 spaces<br \/>   3 spaces<br \/>    4 spaces"},"avatar":"https:\/\/d2k5nsl2zxldvw.cloudfront.net\/images\/avatars\/avatar_2.png","created":"Jun 3 2022 5:50pm PDT","timestamp":1654303855,"automatic":false,"sequenceNumber":1,"actionName":"ADDCOMMENT","shouldShow":true,"clientID":"1654303847866358","reportActionID":"63775"},{"actionName":"CREATED","created":"Jun 3 2022 5:39pm PDT","timestamp":1654303194,"avatar":"https:\/\/d2k5nsl2zxldvw.cloudfront.net\/images\/avatars\/avatar_3.png","message":[{"type":"TEXT","style":"strong","text":"__FAKE__"},{"type":"TEXT","style":"normal","text":" created this report"}],"person":[{"type":"TEXT","style":"strong","text":"__FAKE__"}],"automatic":false,"sequenceNumber":0,"shouldShow":true}];

```

You should have the exact same messages I have.

**Without** the changes from this PR you should see:

<img width="1400" alt="image" src="https://user-images.githubusercontent.com/87341702/171970399-8ad5c5dd-e6dd-4491-8f4c-75769fde612b.png">

With the change (remove [whitespace:pre](https://github.com/Expensify/App/blob/50da22fa41944f42b7c36ba7a91f0ab737eb6cc6/src/styles/styles.js#L138)) from this PR you will see:

<img width="1399" alt="image" src="https://user-images.githubusercontent.com/87341702/171970606-93b5c79c-c98b-4468-8780-cc428590af69.png">

parasharrajat commented:
Sorry, I wasn't able to look into this on the weekend but I will follow up tomorrow. For now, I see that there is a issue with `whitespace:pre` as you mentioned. But I am curious to see that how the whitespaces are preserved without it. Will debug this and let you know.

aldo-expensify commented:
That is exactly an option I was considering, I'll test it today later

On Wed, Jun 8, 2022, 11:19 Rajat Parashar ***@***.***> wrote:

> Looking at the code, I can say that removing whitespace:pre is not
> completely safe. It will not preserve whitespaces in HTML. Your explanation
> is correct #8088 (comment)
> <https://github.com/Expensify/App/issues/8088#issuecomment-1149340794>.
>
> What if we convert the &nbsp; to normal whitespace.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Expensify/App/pull/9195#issuecomment-1150242703>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AU2LVBU5RKQF3Q2TCPRWWHDVODPZZANCNFSM5XCRBU7Q>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

aldo-expensify commented:
I gave it a try to converting the `&nbsp;` to normal spaces, but it looks the same. We still have empty divs taking space.

For the purpose of testing it quicker, I did it in the front here (ReportActionItemFragment):

<img width="782" alt="image" src="https://user-images.githubusercontent.com/87341702/172694637-648b56b0-ab68-448d-aa67-0682a9b5fb61.png">


I also tested adding `enableExperimentalBRCollapsing` and `enableExperimentalGhostLinesPrevention` to `TRenderEngineProvider`, but didn't help.

parasharrajat commented:
Got it. Then backend would be the only viable option.

aldo-expensify commented:
> Got it. Then backend would be the only viable option.

Not necessarily. I'm somewhat confident in that email replies will very often contain `&nbsp`. We could check if they contain that, and if they do don't force the white-space: pre;

And by "backend would be the only viable option", what kind of processing would you suggest? because now we know that just replacing the `&nbsp;` by ' ' won't work

parasharrajat commented:
> because now we know that just replacing the &nbsp; by ' ' won't work

Yeah, it would be too much work. Normally Markup of emails is complex. 

> We could check if they contain that, and if they do don't force the white-space: pre;

Sounds like it will work. Remove the styles from baseFont and apply it to the wrapper.

aldo-expensify committed:
WIP Use flag for comments coming from email reply

A comment flag with source = 'email' should be displayed with
whitespace: 'normal';

aldo-expensify committed:
Remove comment and bg color for debugging

aldo-expensify committed:
Avoid crash if action.originalMessage is missing

This happens when you type a new message and the message
getting rendered is the optimistic data

aldo-expensify committed:
Merge with main, resolve conflicts

aldo-expensify committed:
Apply same logic for email-comment and comment

aldo-expensify commented on src/libs/SelectionScraper/index.js:
```diff
@@ -103,7 +103,7 @@ const replaceNodes = (dom) => {
 
         // Adding a new line after each comment here, because adding after each range is not working for chrome.
-        if (dom.attribs[tagAttribute] === 'comment') {
+        if (dom.attribs[tagAttribute] === 'comment' || dom.attribs[tagAttribute] === 'email-comment') {
```
Maybe we should have a function like 

```js
function isCommentTag(nodeTag) {
    return nodeTag === 'email-comment' || nodeTag === 'comment';
}
````

aldo-expensify commented:
cc @tgolen since you participated in this [convo](https://expensify.slack.com/archives/C01GTK53T8Q/p1654715932206509) and in case you want to have a look

Julesssss commented:
I'll try and remember to check this, but please remind me if I forget to review this once the other PR is merged.

roryabraham commented on src/pages/home/report/ReportActionItemFragment.js:
```diff
@@ -44,6 +44,9 @@ const propTypes = {
     loading: PropTypes.bool,
 
+    /** The reportAction's source */
+    source: PropTypes.string,
```
Should we make this more specific like `PropTypes.oneOf(['email', ...])`?

What other sources could there be?

aldo-expensify replied:
'Chronos': https://github.com/Expensify/Web-Expensify/blob/2f46bbb7f189948b8aac833fed097a9f690446ca/lib/Chronos/Chronos.php#L60, and a few other places
'email': https://github.com/Expensify/Web-Expensify/blob/2f46bbb7f189948b8aac833fed097a9f690446ca/lib/Mailgun/API.php#L1330
'ios', 'android', 'web', 'email': https://github.com/Expensify/Web-Expensify/blob/2f46bbb7f189948b8aac833fed097a9f690446ca/lib/ChatBot/Message.php#L98-L101

Not sure if there could be others... 'Chronos' is not in a constant.

If you use this: `PropTypes.oneOf(['email', ...])`, does it crash if a source doesn't match that list? or will it just warn?

aldo-expensify replied:
<img width="573" alt="image" src="https://user-images.githubusercontent.com/87341702/176305326-ddd5b2e9-2bd9-4ba5-8e77-0da62565531f.png">
Just a warning, no crash, so I think we can safely do that

roryabraham replied:
Seems like a good idea to me, but NAB

aldo-expensify replied:
Changed to: `PropTypes.oneOf(['Chronos', 'email', 'ios', 'android', 'web', 'email', '']),`

roryabraham commented on src/pages/home/report/ReportActionItemFragment.js:
```diff
@@ -86,39 +90,51 @@ const ReportActionItemFragment = (props) => {
+            if (html !== text) {
+                if (props.source === 'email') {
+                    // Messages from email replies usually have complex HTML structure and they don't rely in white-space: pre to preserve spacing,
+                    // instead, the normally use &nbsp;
```
```suggestion
                    // instead, they normally use &nbsp;
```

roryabraham commented on src/pages/home/report/ReportActionItemFragment.js:
```diff
@@ -86,39 +90,51 @@ const ReportActionItemFragment = (props) => {
-            return props.fragment.html !== props.fragment.text
-                ? (
+            if (html !== text) {
+                if (props.source === 'email') {
```
Would be nice to DRY this up a bit:

```js
const editedTag = props.fragment.isEdited ? '<edited></edited>' : '';
const htmlContent = html + editedTag;
return (
    <RenderHTML
        html={props.source === 'email'
            ? `<email-comment>{htmlContent}</email-comment>`
            : `<comment>{htmlContent}</comment>`}
    />
);
```

aldo-expensify replied:
Thanks, that looks much better indeed.

aldo-expensify committed:
Update comment

Co-authored-by: Rory Abraham <47436092+roryabraham@users.noreply.github.com>

aldo-expensify committed:
DRY code

aldo-expensify commented:
Update: DRYer code

roryabraham commented on src/pages/home/report/ReportActionItemFragment.js:
```diff
@@ -44,6 +44,9 @@ const propTypes = {
     loading: PropTypes.bool,
 
+    /** The reportAction's source */
+    source: PropTypes.oneOf(['Chronos', 'email', 'ios', 'android', 'web', 'email', '']),
```
Empty string is a valid source?

roryabraham replied:
And should this be a required prop?

aldo-expensify replied:
I think there are places where we don't pass a source, like here:

https://github.com/Expensify/Web-Expensify/blob/2f46bbb7f189948b8aac833fed097a9f690446ca/api.php#L612-L613

https://github.com/Expensify/Web-Expensify/blob/2f46bbb7f189948b8aac833fed097a9f690446ca/lib/ChatBot/Channel/NativeChat.php#L68

aldo-expensify replied:
> And should this be a required prop?

maybe, I'll look into it. I'll also change the server side PR to always include the source if there is one.

aldo-expensify replied:
I see we have this other usage of `ReportActionItemSingle` here: https://github.com/Expensify/App/blob/de3301c92428e7963bb1bf11ca31d8c49a3511de/src/pages/home/report/ReportActionItemSingle.js#L90-L97

and I'm not passing the `source` there. I'm not familiarized with what are we rendering here..

aldo-expensify replied:
It is rendering the name in the conversation:

<img width="719" alt="image" src="https://user-images.githubusercontent.com/87341702/176321346-9b516317-75d1-4110-b17b-31296273b98f.png">


I personally would keep it as **optional** with default `''` instead of doing it **required** and passing `''` here: https://github.com/Expensify/App/blob/de3301c92428e7963bb1bf11ca31d8c49a3511de/src/pages/home/report/ReportActionItemSingle.js#L90-L97

roryabraham replied:
Okay, this PR LGTM then 👍

aldo-expensify commented:
> I'll try and remember to check this, but please remind me if I forget to review this once the other PR is merged.

@Julesssss I merged the other PR, just letting you know in case you want to review this

Julesssss reviewed (dismissed):
Nice fix, looking good 👍

aldo-expensify commented:
Removed hold for https://github.com/Expensify/Web-Expensify/pull/33993 (it has been deployed to production).. fixing conflicts now.

aldo-expensify committed:
Merge branch 'main' of github.com:Expensify/App into aldo_fix-empty-lines

aldo-expensify commented:
Bump, had to resolve conflicts, need an approval again to get this merged :)

roryabraham commented on src/pages/home/report/ReportActionItemFragment.js:
```diff
@@ -86,39 +90,46 @@ const ReportActionItemFragment = (props) => {
                         )
                 );
             }
+            let {html, text} = props.fragment;
```
IIRC props destructuring is against our React style guide. It's more verbose but we should just reference `props.fragment.html` and `props.fragment.text` instead. Plus it will simplify the diff of this PR

aldo-expensify replied:
We are not destructuring `props` hehe, but `props.fragment`. Do we not allow any kind of destructuring?

roryabraham replied:
I mean, I suppose it's up to interpretation but I think we _do_ allow destructuring, but not when the source of the variable is `props` or `state`. From the [JS style guide](https://github.com/Expensify/App/blob/c67d0d48c050ac0319ad66d38a37596f137844e7/contributingGuides/STYLE.md#destructuring):

> Don't destructure props or state. It makes the source of a given variable unclear. This guideline helps us quickly know which variables are from props, state, or from some other scope.

aldo-expensify replied:
The point of declaring `let html = props.fragment.html` and `let text = props.fragment.text` (or equivalent destructuration) is because a few lines after we may assigned them again:

```js
html = textWithLineBreaks;
text = textWithLineBreaks;
```

I don't think we should do `props.fragment = {...props.fragment, html: textWithLineBreaks, text: textWithLineBreaks};` because we would be modifying the `props` object, and I think this is not advised (I didn't find anything in the  react documentation, but I did find several opinions on that direction).

Would the following change be fine?

```suggestion
            let html = props.fragment.html;
            let text = props.fragment.text;
```

no destructuring, but it is still kind of hiding the source of `html` and `text`... which is the reason why we say we don't want destructuring.

roryabraham replied:
I gotcha

> I don't think we should do props.fragment = {...props.fragment, html: textWithLineBreaks, text: textWithLineBreaks}; because we would be modifying the props object, and I think this is not advised (I didn't find anything in the react documentation, but I did find several opinions on that direction).

This makes sense to me – since we're potentially modifying these variables we can split them off from props. And in that case I don't think it makes any difference whether we using the destructuring syntax or not. Going to resolve this.

Julesssss reviewed (approved):
👍

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/2733888259) to production by @luacmartins in version: 1.1.85-8 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
